### PR TITLE
Feature/notification telegram model

### PR DESCRIPTION
**Pull request by Lex-23**
- create model notification which include models: TelegramNotifier, NotificationType, Notifier
- create notification_manager func
- create telegram_module

### CODE REVIEW

File: cryptostock_api/cryptostock/notification/models.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    account = models.ForeignKey(Account, on_delete=models.CASCADE)
+    type = models.ForeignKey(NotificationType, on_delete=models.CASCADE)
+    telegram_chat_id = models.BigIntegerField(
+        blank=True,
+        null=True,
+        help_text=f"input here 'chat_id' from telegram_bot {os.environ['TG_BOT_URL']}",
+    )
+    email = models.EmailField(
+        blank=True, null=True, help_text="input here email you want for notifications"
+    )
```
Comment (vadim-denisov) [inline cryptostock_api/cryptostock/notification/models.py:61]: So, for every new notifier type se should add new column here?

File: cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    account = Account.objects.get(id=account_id)
+    active_account_notifiers = Notifier.objects.filter(account=account, active=True)
+
+    for notifier in active_account_notifiers:
+        if notifier.type.client.NAME == "Telegram":
+            notifier.type.client.send_notification(
+                func=tg_notify,
+                chat_id=notifier.telegram_chat_id,
+                text=kwargs["tg_text"],
+            ),
```
Comment (vadim-denisov) [inline cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py:16]: This too complex. 
```suggestion
            tg_notify(chat_id=notifier.telegram_chat_id, text=kwargs["tg_text"])
```

If we need `notification_manager` we don't have any effort of `notifier.type.client`

File: cryptostock_api/cryptostock/notification/models.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    """
+    This is a universal model for set up types of users notifications
+    """
+
+    account = models.ForeignKey(Account, on_delete=models.CASCADE)
+    type = models.ForeignKey(NotificationType, on_delete=models.CASCADE)
+    telegram_chat_id = models.BigIntegerField(
+        blank=True,
+        null=True,
+        help_text=f"input here 'chat_id' from telegram_bot {os.environ['TG_BOT_URL']}",
```
Comment (vadim-denisov) [inline cryptostock_api/cryptostock/notification/models.py:57]: If we change {os.environ['TG_BOT_URL']}, we should create migration. Not cool. 
```suggestion
        help_text=f"input here 'chat_id' from telegram_bot",
```

File: cryptostock_api/cryptostock/notification/models.py (hunk 1 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/notification/models.py:61]: Yes. it is weak point. which I don't know how do other methods

File: cryptostock_api/cryptostock/celery_tasks/broker_notification_tasks.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 from account.models import Offer, SalesDashboard
 from celery import shared_task
+from utils.notification_handlers.notification_manager import notification_manager
 
 
 @shared_task(default_retry_delay=10 * 60)
```
Comment (andreyPromaster) [inline cryptostock_api/cryptostock/celery_tasks/broker_notification_tasks.py:6]: Can we move `default_retry_delay` in setting by default?

File: .gitignore [hunk 5]
```diff
@@ -49,7 +49,7 @@ coverage.xml
 *.cover
 *.py,cover
 .hypothesis/
-.pytest_cache/
+cryptostock-api/.pytest_cache/
```
[the remark this answers is not in the document]
Comment (Lex-23) [reply .gitignore:52]: ok. It was automatically chahge

File: cryptostock_api/cryptostock/notification/models.py (hunk 1 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/notification/models.py:61]: maybe make sense to do this through JSONField with **kwargs, I don't know

File: cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py (hunk 2 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py:16]: I do this for universalization for all types of notifications in future.

File: cryptostock_api/cryptostock/celery_tasks/broker_notification_tasks.py (hunk 4 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/celery_tasks/broker_notification_tasks.py:6]: ok, now it is not necessary anyway

File: cryptostock_api/cryptostock/notification/models.py (hunk 3 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/notification/models.py:57]: got it

**Comment by Lex-23** [comment]
[Create telegram client for notifications](https://trello.com/c/EDdquoJK/28-create-telegram-client-for-notifications)

File: cryptostock_api/cryptostock/notification/models.py (hunk 1 above)
Comment (vadim-denisov) [reply cryptostock_api/cryptostock/notification/models.py]: JSONField sounds good. Something like `contacts/consumers_data = JSONField()`

File: cryptostock_api/cryptostock/notification/models.py [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+    @staticmethod
+    def send_notification(func, chat_id, text, **kwargs):
+        """
+        method for notification in telegram
+        :param text: notification text
+        :param chat_id: personal users chat_id
+        :param func: special func for send notification from telegram bot
+        """
+        return func(chat_id, text, **kwargs)
```
Comment (vadim-denisov) [inline cryptostock_api/cryptostock/notification/models.py:24]:
```suggestion
tg_notify = notifier_register("Telegram", notify)
```

File: cryptostock_api/cryptostock/notification/models.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
+from account.models import Account
+from django.db import models
+
+_notifier_storage = {}
+
+
+def notifier_register(cls):
+    _notifier_storage[cls.NAME] = cls
+    return cls
```
Comment (vadim-denisov) [inline cryptostock_api/cryptostock/notification/models.py:9]:
```suggestion
def notifier_register(notify_fn):
    _notifier_storage[cls.NAME] = notify_fn
    return notify_fn
```

File: cryptostock_api/cryptostock/notification/models.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+        if notifier_cls is None:
+            raise ValueError(f"Notifier with name {self.name} doesn't exist")
+        return notifier_cls
+
+    @property
+    def client(self):
+        return self.notifier_cls
+
+    def __str__(self):
+        return self.name
```
Comment (vadim-denisov) [inline cryptostock_api/cryptostock/notification/models.py:42]:
```suggestion
    @property
    def notifier_fn(self):
        notifier_fn = _notifier_storage.get(self.name)
        if notifier_fn is None:
            raise ValueError(f"Notifier with name {self.name} doesn't exist")
        return notifier_fn

    @property
    def send_notification(self, **data):
        return self.notifier_fn(**data)

    def __str__(self):
        return self.name
```

File: cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py (hunk 2 above)
Comment (vadim-denisov) [inline cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py:16]:
```suggestion
    for notifier in active_account_notifiers:
        notifier.type.client.send_notification(notifier, **kwargs["tg_text"])
```

File: cryptostock_api/cryptostock/notification/models.py (hunk 1 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/notification/models.py]: got it

File: cryptostock_api/cryptostock/notification/models.py (hunk 6 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/notification/models.py:24]: good one)

File: cryptostock_api/cryptostock/notification/models.py (hunk 7 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/notification/models.py:9]: got it

File: cryptostock_api/cryptostock/notification/models.py (hunk 8 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/notification/models.py:42]: ok

File: cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py (hunk 2 above)
Comment (Lex-23) [reply cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py:16]: I did better)

Comment (Lex-23) [reply cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py:16]: notifier.type.send_notification(**kwargs)

Comment (Lex-23) [reply cryptostock_api/cryptostock/utils/notification_handlers/notification_manager.py:16]: it works

**Review by vadim-denisov — APPROVED** [review]

**Pull request merged**